text/0000-device-reconnection.md

```diff
@@ -0,0 +1,162 @@
+
+### Protection against connection hijacking
+
+With the possibility of connections migrating from one address to another comes the possibility that malicious actors will attempt to hijack existing connections to redirect them elsewhere (e.g. for amplification attacks). UTP will attempt to protect against hijacks from parties that are not privy to existing traffic. If protection against man-in-the-middle attacks is a concern, using DTLS is expected.
+
+Every message processed by UTP includes an internal session ID that is 16 bits long. This can reasonably be brute-forced on the open internet. As such, the session ID will be modified to be at least 32 bits long, so that it can't easily be guessed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Heartbeats are a very common feature in transport protocols, but considering they are not currently implemented in UTP, their introduction might bring an extra bandwidth overhead. This is especially true for applications that send data very rarely.
+- The implementation extends the `NetworkProtocol` API, which imposes a heavier burden on any future implementations of new protocols in UTP. It could also make it harder to refactor the code (say for protocol stacking).
+- Implementing most of the reconnection decision logic in `NetworkDriver` is the simplest and most straightforward solution, but it does mean that more complexity is added to the part of the UTP code that is already arguably the hardest to understand.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There are two other designs that could have achieved reconnection in a better manner:
+
+- Building UTP over a protocol that already supports seamless mobility (like QUIC; see below). This would have tied UTP to the limitations of that protocol (performance issues, lack of flexibility, limited portability, etc.). It also would have required other services UTP interacts with (like Relay) to support that protocol.
+- Building other protocols on top of the base UDP protocol, which can support seamless mobility. For Relay that would require the server to implement that base UDP protocol, which they might not be able/willing to do. This solution would also prevent users from creating fully encrypted connections (since control messages would be unencrypted).
+
+Not implementing device reconnection means that users will be inconvenienced (by having their online sessions stop working) when roaming from one network to another, which is bad UX. Not providing the feature would also go against our goals of being a top-of-the-class multiplayer solution with a mobile-first philosophy.
+
+# Prior art
+[prior-art]: #prior-art
+
+- The QUIC protocol ([RFC 9600](https://datatracker.ietf.org/doc/html/rfc9000)) is a network protocol built over UDP that automatically handles reconnections the same way we do for the base UDP protocol. (Obviously QUIC is much more complex since it also bundles TLS and congestion control.)
```

Comment:
Typo here. Should read RFC9000 instead of RF9600 the http link is correct though.

Reply:
Fixed. Thanks for catching that.